### PR TITLE
Add type hints for gptneox models

### PR DESCRIPTION
**Pull request by willtai**
# What does this PR do?
Adding missing type hints for `GPTNeoxForCausalLM` and `GPTNeoXModel` as referenced in this issue.(https://github.com/huggingface/transformers/issues/16059#issuecomment-1164898772).

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
[documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
[here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
Anyone in the community please feel free to review😄

@Rocketknight1

### CODE REVIEW

**Review by Rocketknight1 — APPROVED** [review]
Looks great, thank you!

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 1]
```diff
@@ -406,16 +408,16 @@ def set_input_embeddings(self, value):
...
-        attention_mask=None,
-        head_mask=None,
-        inputs_embeds=None,
-        past_key_values=None,
-        use_cache=None,
-        output_attentions=None,
-        output_hidden_states=None,
-        return_dict=None,
-    ):
+        input_ids: Optional[torch.LongTensor] = None,
```
Comment (donelianc) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:411]: According to the [documentation](https://huggingface.co/docs/transformers/model_doc/gpt_neox#transformers.GPTNeoXModel.forward.input_ids) of GPT-NeoX, the parameter `input_ids` is not optional for the class `GPTNeoXModel`.

Comment (Rocketknight1) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py:411]: While this is true, `Optional` in Python types has the very specific meaning that the argument can take the value `None`. See the [typing docs](https://docs.python.org/3/library/typing.html#typing.Optional) for more info on this.

I agree it's not exactly intuitive, but the annotation here is correct as the argument has a default value of `None`!

Comment (donelianc) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py:411]: Great! Thanks for sharing this information. I'm using this PR as an example to contribute to continuing addressing the issue [16059](https://github.com/huggingface/transformers/issues/16059). Again, thanks 😃

**Pull request merged**